api/src/main/java/org/apache/iceberg/SortOrder.java

```diff
@@ -258,14 +257,20 @@ Builder addSortField(String transformAsString, int sourceId, SortDirection direc
     }
 
     public SortOrder build() {
-      if (orderId == 0 && fields.size() != 0) {
+      if (orderId != null && orderId == 0 && fields.size() != 0) {
         throw new IllegalArgumentException("Sort order ID 0 is reserved for unsorted order");
       }
-      if (fields.size() == 0 && orderId != 0) {
+      if (fields.size() == 0 && orderId != null && orderId != 0) {
         throw new IllegalArgumentException("Unsorted order ID must be 0");
       }
 
-      SortOrder sortOrder = new SortOrder(schema, orderId, fields);
+      if (fields.isEmpty()) {
```

Comment:
Could we do this first and then skip all the orderId / size checks? We could probably set order ID back to 1 as well?

Reply:
maybe

if (fields empty)
  if (orderId == 0)
      return empty sort order
  else
       Illegal orderId for empty sort order

if (orderId == 0)
    Illegal orderID for nonEmpty sort order

Reply:
I have two options here:

```
if (fields.isEmpty()) {
  if (orderId == null || orderId == 0) {
    return SortOrder.unsorted();
  } else {
    throw new IllegalArgumentException("Unsorted order ID must be 0");
  }
} else if (orderId != null && orderId == 0) {
  throw new IllegalArgumentException("Sort order ID 0 is reserved for unsorted order");
}
```

or

```
if (fields.isEmpty()) {
  if (orderId != null && orderId != 0) {
    throw new IllegalArgumentException("Unsorted order ID must be 0");
  }
  return SortOrder.unsorted();
}

if (orderId != null && orderId == 0) {
  throw new IllegalArgumentException("Sort order ID 0 is reserved for unsorted order");
}
```

I tend to like the second option a bit more. What do you think, @RussellSpitzer?

Reply:
I think either is good, really I just wanted to have the control flow sorted like you have it now where we can clearly see the branch for unsorted vs sorted code, so beyond that I think it's your preference

Reply:
Agree.

Reply:
Updated. Resolving this.